test/cocotb/osdtestlib/modtestclasses.py

```diff
@@ -0,0 +1,637 @@
+            
+            Args:
+            dut:            device under test.
+            ID:             module address which is assigned to the dut.
+            
+            Returns:
+            -
+        
+        """
+    
+        # setup clock
+        cocotb.fork(Clock(dut.clk, 1000).start())
+    
+        # reset
+        dut._log.info("Resetting DUT")
+        dut.rst <= 1
+
+        dut.trace_valid <= 1
+        dut.debug_out_ready <= 1
+        dut.debug_in.valid <= 1
+    
+        # set module address to the second function parameter
+        dut.id <= ID
+
+        for _ in range(2):
+            yield RisingEdge(dut.clk)
+        dut.rst <= 0
+    
+    
+class Reg_Access:
+    """
+    Class responsible for exchanging messages between debug modules and reading 
+    in their response
+    """
+    
+    PACKET_TYPE_REG                          = 0b00
+    PACKET_TYPE_PLAIN                        = 0b01
+    PACKET_TYPE_EVENT                        = 0b10
+    
+    PACKET_SUBTYPE_REQ_READ_REG_16           = 0b0000
+    PACKET_SUBTYPE_REQ_READ_REG_32           = 0b0001
+    PACKET_SUBTYPE_REQ_READ_REG_64           = 0b0010
+    PACKET_SUBTYPE_REQ_READ_REG_128          = 0b0011
+    PACKET_SUBTYPE_RESP_READ_REG_SUCCESS_16  = 0b1000
+    PACKET_SUBTYPE_RESP_READ_REG_SUCCESS_32  = 0b1001
+    PACKET_SUBTYPE_RESP_READ_REG_SUCCESS_64  = 0b1010
+    PACKET_SUBTYPE_RESP_READ_REG_SUCCESS_128 = 0b1011
+    PACKET_SUBTYPE_RESP_READ_REG_ERROR       = 0b1100
+    PACKET_SUBTYPE_REQ_WRITE_REG_16          = 0b0100
+    PACKET_SUBTYPE_REQ_WRITE_REG_32          = 0b0101
+    PACKET_SUBTYPE_REQ_WRITE_REG_64          = 0b0110
+    PACKET_SUBTYPE_REQ_WRITE_REG_128         = 0b0111
+    PACKET_SUBTYPE_REQ_WRITE_REG_SUCCESS     = 0b1110
+    PACKET_SUBTYPE_REQ_WRITE_REG_ERROR       = 0b1111
+      
+    @cocotb.coroutine    
+    def send_packet(self, dut, packet_driver, packet_utility, destination,
+                    source, packet_type, packet_subtype, message):
+        """function which takes necessary information as its arguments and 
+           transmits a complete packet to a chosen debug module
+        Args:
+            dut:            device under test.
+            destination:    id of the module which is the destination.
+            source:         id of the module which is the source.
+            packet_type:    possible values are: PACKET_TYPE_REG,
+                            PACKET_TYPE_PLAIN and PACKET_TYPE_EVENT.
+            packet_subtype: choose from one of the const values: read or write 
+                            requests of variable length as well as read/write 
+                            responses(success/error).
+            message:        a payload of variable length (up to 8 16 bit words)
+        Returns:
+            -
+        """
+        
+        # build the 3 header flits in advance
+        dest = packet_utility.constructheader1(destination = destination)
```

Comment:
- Why do you pass the packet_utility class instance to the method? You can instead instantiate it directly in this method.
- There's no need to have three method calls to populate the header. Instead, you can have one function construct_header which sets all required fields. Or even better, have a class DiPacket, which represents a single packet in the debug interconnect (DI). You can then set the individual fields in this class and have one method get_flits() for example, which returns all the flits in the packet with the header, etc. set.